libsolutil/LazyInit.h

```diff
@@ -0,0 +1,115 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include <libsolutil/Assertions.h>
+#include <libsolutil/Exceptions.h>
+
+#include <memory>
+#include <optional>
+#include <type_traits>
+#include <utility>
+
+namespace solidity::util
+{
+
+DEV_SIMPLE_EXCEPTION(BadLazyInitAccess);
+
+/**
+ * Represents a lazy-initialized value.
+ * @tparam T the type of the lazy-initialized object; may not be a function, reference, array, or void type; may be const-qualified.
+ */
+template<typename T>
+class LazyInit
+{
+public:
+	using value_type = T;
+
+	static_assert(std::is_object_v<value_type>, "Function, reference, and void types are not supported");
+	static_assert(!std::is_array_v<value_type>, "Array types are not supported.");
+	static_assert(!std::is_volatile_v<value_type>, "Volatile-qualified types are not supported.");
+
+	LazyInit() = default;
+
+	LazyInit(LazyInit const&) = delete;
+
+	// Move constructor must be overridden to ensure that moved-from object is left empty.
+	constexpr LazyInit(LazyInit&& _other) noexcept:
+		m_value(std::move(_other.m_value))
+	{
+		_other.m_value.reset();
+	}
+
+	// Copy and swap idiom.
+	LazyInit& operator=(LazyInit _other) noexcept
+	{
+		this->m_value.swap(_other.m_value);
+	}
+
+	template<typename F>
+	value_type& init(F&& _fun)
```

Comment:
Why not use directly `std::function<T()>` here?

Reply:
``std::function`` has overhead.

Reply:
Do we care? We would get better error messages with this.

Reply:
``std::function`` also doesn't always works for non-copyable captures in lambdas, etc. - I'd say a generic templated callable is the standard way and best way to do this especially since we're in templated context anyways (``std::function``s main benefit is that sometimes it can be used to *avoid* templates, but the whole class is a template anyways, so there's no real benefit in that).

If we want better error messages, we should add a ``static_assert``s with ``std::is_invocable_v<F>`` and ``std::is_convertible_v<std::invoke_result_t<F>, T>`` or something like that.

Reply:
Ok!